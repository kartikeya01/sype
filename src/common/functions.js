export const s4 = () => {
  return Math.floor((1 + Math.random()) * 0x10000)
    .toString(16)
    .substring(1);
};

// generates uniqueID
export const uniqueId = () => {
  return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4();
};

// check singular or plural time for "ago"
export const pluralCheck = s => {
  if (s == 1) {
    return ' ago';
  } else {
    return 's ago';
  }
};

// function to convert timestamp to readable time
export const timeConverter = timestamp => {
  let t = new Date(timestamp * 1000);
  let seconds = Math.floor((new Date() - t) / 1000);

  // check this interval is in years/months/days/hours/minutes?
  // total no of seconds in year are 31536000
  let interval = Math.floor(seconds / 31536000);
  if (interval > 1) {
    return interval + ' year' + pluralCheck(interval);
  }
  // for months
  interval = Math.floor(seconds / 2592000);
  if (interval > 1) {
    return interval + ' month' + pluralCheck(interval);
  }
  // for days
  interval = Math.floor(seconds / 86400);
  if (interval > 1) {
    return interval + ' day' + pluralCheck(interval);
  }
  // for hours
  interval = Math.floor(seconds / 3600);
  if (interval > 1) {
    return interval + ' hour' + pluralCheck(interval);
  }
  // for minutes
  interval = Math.ceil(seconds / 60);
  if (interval > 1) {
    return interval + ' minute' + pluralCheck(interval);
  }
  // for seconds
  return Math.floor(seconds) + ' second' + pluralCheck(seconds);
};

export const timeConverterShortened = timestamp => {
  let t = new Date(timestamp * 1000);
  let seconds = Math.floor((new Date() - t) / 1000);

  // check this interval is in years/months/days/hours/minutes?
  // total no of seconds in year are 31536000
  let interval = Math.floor(seconds / 31536000);
  if (interval > 1) {
    return interval + ' y';
  }

  // for months
  interval = Math.floor(seconds / 2592000);
  if (interval > 1) {
    return interval + ' mo';
  }
  // for days
  interval = Math.floor(seconds / 86400);
  if (interval > 1) {
    return interval + ' d';
  }
  // for hours
  interval = Math.floor(seconds / 3600);
  if (interval > 1) {
    return interval + ' h';
  }
  // for minutes
  interval = Math.ceil(seconds / 60);
  if (interval > 1) {
    return interval + ' m';
  }

  // for seconds
  return Math.floor(seconds) + ' s';
};

export const renderRating = (legitVote, fakeVote) => {
  const totalVotes = legitVote + fakeVote;
  const percentage = Math.floor((legitVote / totalVotes) * 100);

  if (legitVote === 0 && fakeVote === 0) {
    return 'â€¢ ';
  } else if (legitVote === 0 && fakeVote !== 0) {
    return 0;
  } else if (fakeVote === 0 && legitVote !== 0) {
    return 100;
  } else {
    return percentage;
  }
};

export const renderColor = rating => {
  if (rating <= 50) {
    return {color: 'red'};
  } else if (rating >= 80) {
    return {color: 'green'};
  } else if (rating > 50 && rating < 80) {
    return {color: 'orange'};
  } else {
    return {color: 'grey'};
  }
};
