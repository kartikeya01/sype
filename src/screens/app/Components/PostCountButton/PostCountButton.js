import React from 'react';
import {StyleSheet} from 'react-native';
import {Chip} from 'react-native-paper';
import {numberFormat} from '../../../../common/numberFormatter';
import PropTypes from 'prop-types';
import AirbnbPropTypes from 'airbnb-prop-types';

export default class PostCountButton extends React.Component {
  constructor(props) {
    super(props);
  }

  render() {
    const username = this.props.username;

    /** Here, we just need the count of posts,
     * which is created when we query the users.
     */
    let posts_count = this.props.posts_count;

    /** Cheat until we understand proptypes. Because we still need a way to handle string,bool,etc. */
    /** Worst case --> push in 0 for any type other than numeric. */

    /** According to Kartikeya's pull request(7) comment, we should set a miniumum to 0. */
    /** Also note, we should pass in this variable and not the prop if it violates the constraint */
    posts_count =
      Number.isInteger(posts_count) && posts_count >= 0 ? posts_count : 0;

    return <Chip style={styles.chip}>Posts {numberFormat(posts_count)}</Chip>;
  }
}

PostCountButton.propTypes = {
  username: PropTypes.string.isRequired,
  posts_count: AirbnbPropTypes.nonNegativeInteger.isRequired,
};

const styles = StyleSheet.create({
  chip: {
    backgroundColor: 'white',
  },
});
