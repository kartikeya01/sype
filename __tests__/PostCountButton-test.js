import 'react-native';
import React from 'react';
import PostCountButton from '@screens/app/Components/PostCountButton/PostCountButton';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import {Chip} from 'react-native-paper';
import {numberFormat} from '../src/common/numberFormatter';
import checkPropTypes from 'check-prop-types';
var assert = require('assert');

/** ================ HELPER FUNCTIONS =============== */

/** Notice, always forcing the other props to be correct, to not produce warnings and test each prop individually. */
function testAllProperPropsPassed(username, posts_count) {
  it('Username value is the PostCount Button prop value', () => {
    const wrapper = shallow(
      <PostCountButton username={username} posts_count={0} />,
    );
    const instance = wrapper.instance();
    const usernameProp = instance.props.username;
    expect(username).toEqual(usernameProp);
  });

  it('Posts Count value is the PostCount Button prop value', () => {
    const wrapper = shallow(
      <PostCountButton username={''} posts_count={posts_count} />,
    );
    const instance = wrapper.instance();
    const postsCountProp = instance.props.posts_count;
    expect(posts_count).toEqual(postsCountProp);
  });
}

/** Pass in all props for when we require them */
function testChipInnerText(username, posts_count) {
  it('Posts Chip should Display Posts ' + posts_count, () => {
    const wrapper = shallow(
      <PostCountButton username={username} posts_count={posts_count} />,
    );
    const chipWrapper = wrapper.find(Chip);
    const chipText = chipWrapper.text();
    /**Check if posts_count is undefined -> expect 0 */
    const expectedPostsCount = posts_count === undefined ? 0 : posts_count;

    /**Formatting seems to work with ''  */
    expect(chipText).toEqual('Posts ' + numberFormat(expectedPostsCount));
  });
}

/** ================ SNAPSHOT TEST ================ */

/** NOTE, snapshot tests will give warnings with no required props passed. */
test('Post Count Button Snapshot Test', () => {
  const tree = renderer
    .create(<PostCountButton username={''} posts_count={0} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

/** ================ UNIT TESTS ================ */
describe('PostCountButton Unit Testing', () => {
  const integerPostCountTest = 1;
  const stringUsernameTest = 'bsefa';
  testAllProperPropsPassed(stringUsernameTest, integerPostCountTest);
  testChipInnerText(stringUsernameTest, integerPostCountTest);

  it('Throws failed propType erorr for not providing any props', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {},
      'prop',
      PostCountButton,
    );
    assert(
      result.includes(
        'Failed prop type: The prop `username` is marked as required',
      ),
    );
  });

  it('Throws failed propType erorr for not providing username', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {posts_count: 1},
      'prop',
      PostCountButton,
    );
    assert(
      result.includes(
        'Failed prop type: The prop `username` is marked as required',
      ),
    );
  });

  it('Throws failed propType erorr for not providing posts_count', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: 'bujar'},
      'prop',
      PostCountButton,
    );
    /** console.log(result);
     * This actually prints: Failed prop type: posts_count in function PostCountButton {
     *   (0, _classCallCheck2.default)(this, PostCountButton);
     *   return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(PostCountButton).call(this, props));
     * } must be an integer
     * But incase (becasue I don't necessarily know how it all works) the inner function changes, we don't want to expect what happens inside. Just that its the right class and right type.
     */
    /** NOTICE, that this does not check for posts_count to be require or non-negative! Which is the actual restriction to the prop type */
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be an integer'),
    );
  });

  it('Throws Failed propType Error for providing username value of undefined and not required value', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: undefined},
      'prop',
      PostCountButton,
    );
    assert(
      result.includes(
        'Failed prop type: The prop `username` is marked as required in `function PostCountButton',
      ) && result.includes('but its value is `undefined`.'),
    );
  });

  it('Throws Failed propType Error for providing username value of true and not string', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: true},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: Invalid prop `username` of type `boolean` supplied to `function PostCountButton',
      ) && result.includes('expected `string`.'),
    );
  });

  it('Throws Failed propType Error for providing username value of false and not string', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: false},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: Invalid prop `username` of type `boolean` supplied to `function PostCountButton',
      ) && result.includes('expected `string`.'),
    );
  });

  it('Throws Failed propType Error for providing username value of integer number and not string', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: 12},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: Invalid prop `username` of type `number` supplied to `function PostCountButton',
      ) && result.includes('expected `string`.'),
    );
  });

  it('Throws Failed propType Error for providing username value of double number and not string', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: 12.1},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: Invalid prop `username` of type `number` supplied to `function PostCountButton',
      ) && result.includes('expected `string`.'),
    );
  });

  it('Throws Failed propType Error for providing username value of array and not string', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: []},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: Invalid prop `username` of type `array` supplied to `function PostCountButton',
      ) && result.includes('expected `string`.'),
    );
  });

  it('Throws Failed propType Error for providing username value of object and not string', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: {}},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: Invalid prop `username` of type `object` supplied to `function PostCountButton',
      ) && result.includes('expected `string`.'),
    );
  });

  it('Throws Failed propType Error for providing posts_count value of undefined and not required value', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: undefined},
      'prop',
      PostCountButton,
    );
    /** This doesn't compare the type. Just realizes that it is not an integer.
     * Different from PropTypes logs...
     * Will this ultimately matter?
     */
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be an integer'),
    );
  });

  it('Throws Failed propType Error for providing posts_count value of true and must be an integer', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: true},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be an integer'),
    );
  });

  it('Throws Failed propType Error for providing posts_count value of false and must be an integer', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: false},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be an integer'),
    );
  });

  it('Throws Failed propType Error for providing posts_count value of number and must be an integer', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: 12.1},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be an integer'),
    );
  });

  it('Throws Failed propType Error for providing posts_count value of array and must be an integer', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: []},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be an integer'),
    );
  });

  it('Throws Failed propType Error for providing posts_count value of object and must be an integer', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: {}},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be an integer'),
    );
  });

  it('Throws Failed propType Error for providing posts_count value of negative number and must be an non-negative integer', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: -1},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be a non-negative number'),
    );
  });

  it('Throws Failed propType Error for providing posts_count value of negative double number and must be an integer', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: -1.1},
      'prop',
      PostCountButton,
    );
    //console.log(result);

    /** In airbnb documentaion, a non-negative integer is just defined as an integer and non negative number
     * Because integer case fails, it returns must be an integer!
     */
    assert(
      result.includes(
        'Failed prop type: posts_count in function PostCountButton',
      ) && result.includes('must be an integer'),
    );
  });

  it('Undefined returned for username and profile_image string type', () => {
    let result = checkPropTypes(
      PostCountButton.propTypes,
      {username: '', posts_count: 0},
      'prop',
      PostCountButton,
    );
    //console.log(result);
    //result returns undefined when the values are true, as they should be.
    assert(result === undefined);
  });
});
