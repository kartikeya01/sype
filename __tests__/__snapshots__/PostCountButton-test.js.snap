// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Count Button Snapshot Test 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderColor": "white",
      "borderRadius": 16,
      "borderStyle": "solid",
      "borderWidth": 0.5,
      "elevation": 0,
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityStates={Array []}
    accessible={true}
    focusable={false}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 16,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "paddingHorizontal": 4,
        }
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
              "textAlign": "left",
            },
            Array [
              Object {
                "lineHeight": 24,
                "marginVertical": 4,
                "minHeight": 24,
                "textAlignVertical": "center",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "marginLeft": 8,
                "marginRight": 8,
              },
              undefined,
            ],
          ]
        }
      >
        Posts 
        0
      </Text>
    </View>
  </View>
</View>
`;
